server/fleet/app.go

```diff
@@ -239,6 +258,29 @@ func (c *AppConfig) ApplyDefaults() {
 	c.WebhookSettings.Interval.Duration = 24 * time.Hour
 }
 
+// EnableStrictDecoding enables strict decoding of the AppConfig struct.
+func (c *AppConfig) EnableStrictDecoding() { c.strictDecoding = true }
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (c *AppConfig) UnmarshalJSON(b []byte) error {
+	// Decode any legacy settings that might be present.
+	var legacy legacyConfig
+	if err := json.Unmarshal(b, &legacy); err != nil {
+		return err
+	}
+
+	// TODO(roperzh): define and assign legacy settings to new fields. E.g:
+	// c.Features = legacy.HostSettings
+
+	// Decode the config, overriding any legacy settings.
+	type alias *AppConfig
+	decoder := json.NewDecoder(bytes.NewReader(b))
+	if c.strictDecoding {
+		decoder.DisallowUnknownFields()
+	}
+	return decoder.Decode(alias(c))
```

Comment:
The problem with `json.Decoder` is that it consumes a stream of JSON values, so when it's done decoding one value, it is not guaranteed that it consumed the whole input (e.g. https://pkg.go.dev/encoding/json#example-Decoder). What we should do to enforce that is to check that it returns `io.EOF` (and eat that error as a success).

Reply:
lol, Michal mentioned this a little while ago, and yesterday I found [this](https://github.com/golang/go/issues/36225) by accident, but I still failed.

Fix pushed.